### PR TITLE
Add PrioritizedIntents to support multiple shortcut configurations

### LINKED ISSUE #43694: [web] Add support for spacebar Shortcut to include ScrollIntent
I am building web pages with Flutter Web and the widget layout is as follows.
```
Scaffold(
   body: SingleChildScrollView(
      child: Column(
          children: <Widget>[
              // page widgets
```

All of my widgets that build the content of the page are the children of the column. Scrolling works fine if I use my trackpad or click on the page and drag (which makes sense because this is how mobile applications scroll). I understand it is currently working as it was designed for mobile. However, there is no scroll bar on the side of the page as there normally would be with any other website. Since we can only make webpages scrollable by using nested scrollable widgets like SingleChildScrollView, there should be a way to recognize nested SingleChildScrollViews that hold the body of the webpage as being scrollable with a website scroll bar so that functions such as scrolling by pressing spacebar, page up/page down, up/down arrow keys or by using a side scroll bar are possible.

### PR DESCRIPTION
**Pull request by Piinks**
## Description

This adds a new Intent - `PrioritizedIntents`
This will allow for multiple Intents to be associated with a single `LogicalKeySet`, so multiple Shortcuts can be evaluated.

For example - the web platform expects that, in the right context, the spacebar will scroll a scroll view. This has not been possible because the spacebar also serves as a shortcut to activate switches and buttons.

In a `PrioritizedAction`, the listed `Intents` will be evaluated such that the right `Action` will be invoked. In our web case, the `ActivateIntent` will be evaluated first, followed by the `ScrollIntent`. So if there were a `Checkbox` inside of a `ScrollView`, the space bar shortcut could scroll the view or activate the checkbox, depending on the location of the `primaryFocus`. The test case here shows this example.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/43694

## Tests

Shortcuts support multiple intents

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Piinks** [comment]
@gspencergoog perhaps for discussing after the holidays, this is what I had in mind for the multi-Intent we mentioned earlier this week. :)

**Comment by gspencergoog** [comment]
@Piinks This is super cool! I say write the test and ship it!

File: packages/flutter/lib/src/widgets/actions.dart [hunk 1]
```diff
@@ -1321,3 +1321,50 @@ class DismissIntent extends Intent {
...
+  /// progression through the ordered intents stops.
+  final List<Intent> orderedIntents;
+}
+
+/// An [Action] that iterates through a list of [Intent]s, invoking the first
+/// that is enabled.
+class PrioritizedAction extends Action<PrioritizedIntents> {
+  late Action<dynamic> _selectedAction;
+  late Intent _selectedIntent;
+
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/actions.dart:1344]:
```suggestion

```

File: packages/flutter/lib/src/widgets/actions.dart [hunk 2]
```diff
@@ -1321,3 +1321,50 @@ class DismissIntent extends Intent {
...
+      );
+      if (candidateAction != null && candidateAction.isEnabled(candidateIntent)) {
+        _selectedAction = candidateAction;
+        _selectedIntent = candidateIntent;
+        return true;
+      }
+    }
+    return false;
+  }
+
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/actions.dart:1363]:
```suggestion

```

File: packages/flutter/lib/src/widgets/actions.dart [hunk 3]
```diff
@@ -1321,3 +1321,48 @@ class DismissIntent extends Intent {
...
+  final List<Intent> orderedIntents;
+}
+
+/// An [Action] that iterates through a list of [Intent]s, invoking the first
+/// that is enabled.
+class PrioritizedAction extends Action<PrioritizedIntents> {
+  late Action<dynamic> _selectedAction;
+  late Intent _selectedIntent;
+  @override
+  bool isEnabled(PrioritizedIntents intent) {
```
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/actions.dart:1345]: tiny nit: perhaps add blank lines before and after the methods?

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/lib/src/widgets/actions.dart [hunk 4]
```diff
@@ -1321,3 +1321,48 @@ class DismissIntent extends Intent {
...
+/// An [Intent] that evaluates a series of specified [orderedIntents] for
+/// execution.
+class PrioritizedIntents extends Intent {
+  /// Creates a set of const [PrioritizedIntents].
+  const PrioritizedIntents({
+    required this.orderedIntents,
+  })  : assert(orderedIntents != null);
+
+  /// List of intents to be evaluated in order for execution. When an
+  /// [Intent.isEnabled] returns true, that intent will be invoked and
```
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/actions.dart:1334]: Do you mean `Action.isEnabled` here?

Comment (Piinks) [reply packages/flutter/lib/src/widgets/actions.dart:1334]: Ah yes! Thank you!

Comment (Piinks) [inline packages/flutter/lib/src/widgets/actions.dart:1334]:
```suggestion
  /// [Action.isEnabled] returns true, that action will be invoked and
```

File: packages/flutter/lib/src/widgets/actions.dart [hunk 5]
```diff
@@ -1321,3 +1321,48 @@ class DismissIntent extends Intent {
...
+  /// progression through the ordered intents stops.
+  final List<Intent> orderedIntents;
+}
+
+/// An [Action] that iterates through a list of [Intent]s, invoking the first
+/// that is enabled.
+class PrioritizedAction extends Action<PrioritizedIntents> {
+  late Action<dynamic> _selectedAction;
+  late Intent _selectedIntent;
+  @override
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/actions.dart:1344]:
```suggestion

  @override
```

File: packages/flutter/lib/src/widgets/actions.dart [hunk 6]
```diff
@@ -1321,3 +1321,48 @@ class DismissIntent extends Intent {
...
+      );
+      if (candidateAction != null && candidateAction.isEnabled(candidateIntent)) {
+        _selectedAction = candidateAction;
+        _selectedIntent = candidateIntent;
+        return true;
+      }
+    }
+    return false;
+  }
+  @override
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/actions.dart:1362]:
```suggestion

  @override
```

**Comment by devanflores** [comment]
@Piinks Is there any way to disable this functionality through an option?

**Comment by Piinks** [comment]
> @Piinks Is there any way to disable this functionality through an option e.g. prevent spacebar from scrolling with a boolean?

Hey @devanflores you can provide your own shortcuts to `MaterialApp.shortcuts` to override the default shortcuts.
I think in the case you mentioned in #43694, you might also be able to put a Shortcut around your TextField that will override the top level Shortcuts. I have not tried it out yet.

**Comment by gspencergoog** [comment]
@devanflores you can override any shortcut binding by placing a "Shortcuts" widget above the widgets you would like to override it for (or at the top of your app, as a child of the MaterialApp/CupertinoApp widget), defining a different shortcut binding for the same keys. You can also disable it entirely by binding it to a `DoNothingIntent()`.

So, like this:

```
Widget build(BuildContext context) {
  return Shortcuts(
    shortcuts: <LogicalKeySet, Intent> {
      LogicalKeySet(LogicalKeyboardKey.space): DoNothingIntent(),
    },
	child: child,
  );
}
```

**Pull request merged**